pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -131,9 +131,13 @@ def attach_datamodule(self, model, datamodule: Optional[LightningDataModule], st
             if is_overridden('predict_dataloader', datamodule):
                 model.predict_dataloader = datamodule.predict_dataloader
 
-            # Override transfer_batch_to_device if dataset-specific to_device logic has been defined in datamodule
+            # Override data transfer hooks if dataset-specific to_device logic has been defined in datamodule
+            if is_overridden('on_before_batch_transfer', datamodule):
+                model.on_before_batch_transfer = datamodule.on_before_batch_transfer
             if is_overridden('transfer_batch_to_device', datamodule):
                 model.transfer_batch_to_device = datamodule.transfer_batch_to_device
+            if is_overridden('on_after_batch_transfer', datamodule):
+                model.on_after_batch_transfer = datamodule.on_after_batch_transfer
 
             self.trainer.datamodule = datamodule
             datamodule.trainer = self.trainer
```

Comment:
I know you didn't "invent this" but I would like to comment anyway that patching the model's methods like this is bad and we should make an effort to avoid this in the future. The same is done with data loaders. The user doesn't really know this, we are essentially destroying the definition of the user's model. We also never reset it, i.e., if the user want's to perform some operations with the model after fit, it will misbehave.
I will finally open an issue for this :)

Reply:
#6030

Reply:
agreed.